compiler/sem/injectdestructors.nim

```diff
@@ -667,118 +598,57 @@ proc expandAsgn(tree: MirTree, ctx: AnalyseCtx, ar: AnalysisResults,
-        c.changeTree(tree, operator): MirNode(kind: mnkMove)
       else:
-        # no hook call nor destructive move is required
-        discard "nothing to do"
-
-  else:
-    # a move is not possible -> copy
-    c.replaceMulti(tree, stmt, bu):
-      # copies to locals or globals can't introduce cyclic structures, as
-      # those are standlone and not part of any other structure
-      let maybeCyclic =
-        tree[dest].kind notin {mnkLocal, mnkTemp, mnkParam, mnkGlobal}
-      let
-        a = bu.bindMut(tree, dest)
-        b = bu.inline(tree, source)
-
-      genCopy(bu, ctx.graph, env, a, b, maybeCyclic)
-
-proc expandDef(tree: MirTree, ctx: AnalyseCtx, ar: AnalysisResults,
-               env: var MirEnv, at: NodePosition, pos: InstrPos,
-               c: var Changeset) =
-  ## Depending on whether the source can be moved out of, either rewrites the
-  ## 'def' at `at` into a call to the ``=copy`` hook call or into a
-  ## destructive or non-destructive move. `pos` is the data-flow instruction.
-  let
-    dest     = tree.child(at, 0)
-    operator = tree.child(at, 1)
-    source   = tree.child(operator, 0)
-  case isMove(tree, ar.moves[], operator)
-  of false:
-    # a copy is required. Transform ``def x = copy a.b`` into:
-    #   def x
-    #   bind _1 = a.b
-    #   =copy(name x, arg _1)
-    c.replace(tree, operator): MirNode(kind: mnkNone)
-    c.insert(tree, tree.sibling(at), source, bu):
-      let
-        a = bu.bindMut(tree, dest)
-        b = bu.inline(tree, source)
-      # the destination can only be a cell-like location (local, global,
-      # etc.), no cycle can possibly be introduced
-      genCopy(bu, ctx.graph, env, a, b, false)
-  of true:
-    assert tree[operator].kind == mnkSink
-    if needsReset(tree, ctx.cfg, ar, computePath(tree, source), pos):
-      # the value can be moved, but the location needs to be reset. Transform
-      # ``def x = sink a.b`` into:
-      #   bind_mut _1 = a.b
-      #   def x = move _1
-      #   wasMoved(name x)
-      var tmp, clear: Value
-      c.insert(tree, at, source, bu):
-        (tmp, clear) = bu.destructiveMoveOperands(tree, source)
-      c.replaceMulti(tree, operator, bu):
-        bu.move tmp
-      c.insert(tree, tree.sibling(at), source, bu):
-        genWasMoved(bu, ctx.graph, clear)
+        # the value can be moved without the source location having to be
+        # cleared
+        c.changeTree(tree, operator):
+          MirNode(kind: mnkMove, typ: tree[operator].typ)
     else:
-      # turn into a ``Move`` operation
+      # the value cannot be moved, turn the sink into a copy
       c.changeTree(tree, operator):
-        MirNode(kind: mnkMove, typ: tree[operator].typ)
+        MirNode(kind: mnkCopy, typ: tree[operator].typ)
 
+    if tree[stmt].kind == mnkAsgn and not isAlive:
+      # the assignment initializes the location
+      c.changeTree(tree, stmt): MirNode(kind: mnkInit)
+  else:
+    # it's a move or move already, so nothing to change there
```

Comment:
```suggestion
    # it's a move or destructive move already, so nothing to change there
```

Not sure if 'destructive' is what you meant, but IIRC, mir/proto-mir can generate either if safe, and will do so instead of emiting a `sink` for injectdestructors to resolve. That could be the case if the destination had no value to begin with or perhaps the location was reset at that point (`isAlive == false`).

Reply:
Ah, good catch -- a copy & past mistake. I meant to use "copy" here.

Reply:
```suggestion
    # it's a move or copy already, so nothing to change there
```